### PR TITLE
More efficient median binned stat calculation and new min & max binned stats

### LINKED ISSUE #5100: inefficient median calculation in scipy.stats.binned_statistic_dd
The binned_statistic functions are excellent - a great resource for reducing large geospatial point datasets.  I've found that the binned_statistic_2d is very efficient when the stat is mean, std, count or sum.  
However, the median calculation is very inefficient.  
I glanced at the source, and I believe the median and the generic statistic functions are being computed for every single bin, regardless of whether those bins actually contain points.  It seems like it should be trivial to limit these stats to bins with count > 0.  
It would also be nice to have efficient 'min' and 'max'  stats available as options, instead of the user passing as a function.

### PR DESCRIPTION
**Pull request by samueljackson92**
- Prevent calculating the median on bins where count == 0.
- Add min/max options to `binned_statistic` function
- Closes #5100

### CODE REVIEW

File: scipy/stats/_binned_statistic.py [hunk 1]
```diff
@@ -280,6 +280,10 @@ def binned_statistic_dd(sample, values, statistic='mean',
             referenced.
           * 'sum' : compute the sum of values for points within each bin.
             This is identical to a weighted histogram.
+          * 'max' : compute the maximum value for points within each bin.
+            Empty bins will be represented by NaN.
+          * 'min' : compute the minimum value for points within each bin.
+            Empty bins will be represented by NaN.
```
Comment (ev-br) [inline scipy/stats/_binned_statistic.py]: These new options should also be listed in the docstrings of `binned_statistic_2d` and `binned_statistic`.

**Comment by ev-br** [comment]
Thanks for the contribution and sorry it received no attention so far.

Can you (or @dshean) demonstrate a performance improvement for the median calculation on some data? I briefly played with this PR and did not see much of an improvement on a synthetic dataset.

**Comment by juliantaylor** [comment]
maybe I'm dense but isn't it impossible for the value array to have size zero?

I don't think the median is the actual bottleneck but there is a simple optimization, as its working on a copy one can use `overwrite_input=True`

**Comment by ev-br** [comment]
This does not apply cleanly anymore after a recent enhancement to binned_statistic.py.  I guess it's easier to redo than rebase.  I'm going to close this with a *needs-champion* label. @samueljackson92 are you interested in finishing this up? In that case please resubmit.

**Comment by ahalota** [comment]
@samueljackson92 MAINT: Limit median computation to bins where count > 0 a80c58e
@samueljackson92 ENH: Add min and max binned statistics. 80386a7

Hi, 

Looking through these two commits, should the same change you made to median also be applied to min/max, i.e. Don't bother calculating min/max if the count is 0?


I can't tell if this initial pull request was ever merged or not (not too familiar with github UI)... I'm definitely noticing some efficiency issues on 'min' as added by @thomaspingel in #6911, in the case of a sparse input array (10% of bins filled)

**Comment by tylerjereddy** [comment]
> I can't tell if this initial pull request was ever merged or not 

This specific one was closed, not merged. There should be a red "closed" label in the top left corner of the UI.

**Pull request closed without merging**